contracts/staking/test/stake_test.ts

```diff
@@ -1,47 +1,357 @@
+        });
+        it('should fail to unstake after being inactive for <1 epoch', async () => {
+            const amount = StakingWrapper.toBaseUnitAmount(10);
+            await staker.stakeAsync(amount);
+            await staker.moveStakeAsync({ state: StakeState.Active }, { state: StakeState.Inactive }, amount);
+            await staker.goToNextEpochAsync();
+            await staker.unstakeAsync(amount, new StringRevertError('INSUFFICIENT_FUNDS'));
+        });
+        it('should fail to unstake in same epoch that inactive/withdrawable stake has been reactivated', async () => {
+            const amount = StakingWrapper.toBaseUnitAmount(10);
+            await staker.stakeAsync(amount);
+            await staker.moveStakeAsync({ state: StakeState.Active }, { state: StakeState.Inactive }, amount);
+            await staker.goToNextEpochAsync(); // stake is now inactive
+            await staker.goToNextEpochAsync(); // stake is now withdrawable
+            await staker.moveStakeAsync({ state: StakeState.Inactive }, { state: StakeState.Active }, amount);
+            await staker.unstakeAsync(amount, new StringRevertError('INSUFFICIENT_FUNDS'));
+        });
+        it('should fail to unstake one epoch after inactive/withdrawable stake has been reactivated', async () => {
+            const amount = StakingWrapper.toBaseUnitAmount(10);
+            await staker.stakeAsync(amount);
+            await staker.moveStakeAsync({ state: StakeState.Active }, { state: StakeState.Inactive }, amount);
+            await staker.goToNextEpochAsync(); // stake is now inactive
+            await staker.goToNextEpochAsync(); // stake is now withdrawable
+            await staker.moveStakeAsync({ state: StakeState.Inactive }, { state: StakeState.Active }, amount);
+            await staker.goToNextEpochAsync(); // stake is active and not withdrawable
+            await staker.unstakeAsync(amount, new StringRevertError('INSUFFICIENT_FUNDS'));
+        });
+        it('should fail to unstake >1 epoch after inactive/withdrawable stake has been reactivated', async () => {
+            const amount = StakingWrapper.toBaseUnitAmount(10);
+            await staker.stakeAsync(amount);
+            await staker.moveStakeAsync({ state: StakeState.Active }, { state: StakeState.Inactive }, amount);
+            await staker.goToNextEpochAsync(); // stake is now inactive
+            await staker.goToNextEpochAsync(); // stake is now withdrawable
+            await staker.moveStakeAsync({ state: StakeState.Inactive }, { state: StakeState.Active }, amount);
+            await staker.goToNextEpochAsync(); // stake is active and not withdrawable
+            await staker.goToNextEpochAsync(); // stake is active and not withdrawable
+            await staker.unstakeAsync(amount, new StringRevertError('INSUFFICIENT_FUNDS'));
+        });
+    });
+    describe('Simulations', () => {
+        it('Simulation from Staking Spec', async () => {
```

Comment:
It would be nice to have a simulation that is a bit shorter for debugging purposes. If this simulation finds a bug, it may be a bit non-trivial to figure out where the bug lives.

Reply:
That's a good point. I worked to include all of the edge cases individually, in the tests above. I wanted to keep this just to have some semblance of a legitimate sequence of interactions.

Reply:
Fair enough! Before writing this comment, I had the thought that the Simulations would likely not get hit in the case of a failure (because of an earlier failure). Up to you about whether to include a shorter sequence. 

Out of scope: Given the complexity of the staking logic, I think that some potential bugs could be found in pathologically long sequences (if there are any). I think it would really behoove us to do some fuzzing using the created actors.